<launch>
	<arg name="marker_size" default="19.4" />
	<arg name="max_new_marker_error" default="0.08" />
	<arg name="max_track_error" default="0.2" />
	<arg name="cam_image_topic" default="/camera/rgb/image_color" />
    <arg name="cam_info_topic" default="/camera/rgb/camera_info" />
	<arg name="output_frame" default="/camera_rgb_frame" />
	<node name="ar_track_alvar" pkg="ar_track_alvar" type="individualMarkersNoKinect" respawn="false" output="screen" args="$(arg marker_size) $(arg max_new_marker_error) $(arg max_track_error) $(arg cam_image_topic) $(arg cam_info_topic) $(arg output_frame)" />


    <!-- joy node -->
    <node respawn="false" pkg="joy" type="joy_node" name="joy" >
        <param name="dev" type="string" value="/dev/input/js0" />
        <!-- <param name="deadzone" value="0.12" /> -->
        <param name="autorepeat_rate" value="10.0" />
    </node>

    <!-- Axes -->

    <node pkg="vrep_ros_teleop" type="teleop.py" name="teleop">
        <param name="axis_linear" value="1" type="int"/>
        <param name="axis_angular" value="0" type="int"/>
        <param name="scale_linear" value="0.3" type="double"/>
        <param name="scale_angular" value="-1.0" type="double"/>
        <remap from="/teleop/joy" to="/joy"/>
    </node>

    <node pkg="topic_tools" type="mux" name="cmd_mux"
          args="/cmd_vel_mux/input/navi /teleop/twistCommand /mux/autoCommand">
    </node>

    <node pkg="vrep_ros_teleop" type="teleop_mux.py" name="teleop_mux">
        <param name="joystick_button" value="0" type="int"/>
        <param name="joystick_topic" value="/teleop/twistCommand"/>
        <param name="auto_button" value="1" type="int"/>
        <param name="auto_topic" value="/mux/autoCommand"/>
        <remap from="/teleop_mux/joy" to="/joy"/>
    </node>

    <node pkg="ar_slam" type="turtle_slam_angle.py" name="slam" output="screen">
        <param name="ignore_id" value="False"/>
        <param name="target_frame" value="/map"/>
        <param name="odom_frame" value="/odom"/>
        <param name="body_frame" value="/base_link"/>
        <!-- Expected precision of wheel displacement measurement (in meter) -->
        <param name="position_uncertainty" value="0.01"/>
        <param name="angular_uncertainty" value="0.01"/>
        <!-- Expected precision of ar_tag localisation (in meter) -->
        <param name="ar_precision" value="0.10"/>
        <!-- Initial value of X,Y,Theta -->
        <param name="initial_x" value="0.0"/>
        <param name="initial_y" value="0.0"/>
        <param name="initial_theta" value="0.0"/>
    </node>

    <!-- Layered Costmap -->

    <!--<node name="costmap_node" pkg="costmap_2d" type="costmap_2d_node" >-->
        <!--<rosparam file="/home/turtlebot/catkin_ws/src/ar_slam/launch/2D_costmap.yaml" command="load" ns="costmap" />-->
    <!--</node>-->

    <node pkg="costmap_2d" type="costmap_2d_node" name="costmap" output="screen">
        <rosparam file="$(find turtlebot_navigation)/param/costmap_common_params.yaml" command="load" ns="costmap/obstacle_layer" />
        <param name="costmap/global_frame" value="map"/>
        <param name="costmap/robot_base_frame" value="base_link"/>
        <param name="costmap/update_frequency" value="3.0"/>
        <param name="costmap/publish_frequency" value="0.25"/>
        <param name="costmap/static_map" value="false"/>
        <param name="costmap/rolling_window" value="false"/>
        <param name="costmap/obstacle_layer/raytrace_range" value="2.5"/>
        <!--<param name="costmap/width" value="60"/>-->
        <!--<param name="costmap/height" value="60"/>-->
        <!--<param name="costmap/resolution" value="0.05"/>-->
        <param name="transform_tolerance" value="0.5"/>
        <param name="costmap/publish_voxel_grid" value="true"/>
        <!--<param name="costmap/origin_x" value="-10"/>-->
        <!--<param name="costmap/origin_y" value="-40"/>-->
        <param name="costmap/unknown_cost_value" value="-1"/>
        <param name="costmap/track_unknown_space" value="true"/>

        <param name="costmap/robot_radius" value="0.20"/>
        <param name="costmap/footprint" value="[[0.2,0.2],[0.2,-0.2],[-0.2,-0.2],[-0.2,0.2]]"/>
        <param name="costmap/width" value="30"/>
        <param name="costmap/height" value="30"/>
        <param name="costmap/origin_x" value="0"/>
        <param name="costmap/origin_y" value="0"/>
        <param name="costmap/resolution" value="0.05"/>
        <param name="costmap/obstacle_layer/observation_sources" value="klaser"/>
        <param name="costmap/obstacle_layer/track_unknown_space" value="true"/>
        <param name="costmap/obstacle_layer/unknown_cost_value" value="-1"/>
        <param name="costmap/obstacle_layer/klaser/topic" value="/scan"/>
        <param name="costmap/obstacle_layer/klaser/data_type" value="LaserScan"/>
        <param name="costmap/obstacle_layer/klaser/clearing" value="true"/>
        <param name="costmap/obstacle_layer/klaser/marking" value="true"/>
        <param name="costmap/obstacle_layer/klaser/obstacle_range" value="5.0"/>
        <param name="costmap/obstacle_layer/klaser/raytrace_range" value="5.0"/>

    </node>

</launch>
